Proposed IOUring changes

njhill opened the pull request:
Motivation

After reviewing the latest IOUring work I made a few changes to illustrate some of my prior review suggestions (plus some other things). This is not tested and I expect the changes will need some discussion/agreement!

Modifications

- Reserve an entry in the SQE array for the eventfd SQE so that we don't write it every time
- Don't pollin for eventfd, just read (I think this is ok due to the FAST_POLL magic)
- Combine submission/waiting, don't submit until either the queue is full or there's no more work to do (when entering wait)
- Rework event loop sequencing w.r.t. checking for tasks/completions
- Change some longs to back to ints where I don't think they need to be converted
- Track head/tail and masks (constant) as instance vars to avoid reading from the mapped buffer every time

Result

Hopefully further perf gains

njhill committed:
Proposed IOUring changes

Motivation

After reviewing the latest IOUring work I made a few changes to
illustrate some of my prior review suggestions (plus some other things).
This is not tested and I expect the changes will need some
discussion/agreement!

Modifications

- Reserve an entry in the SQE array for the eventfd SQE so that we don't
write it every time
- Don't pollin for eventfd, just read (I think this is ok due to the
FAST_POLL magic)
- Combine submission/waiting, don't submit until either the queue is
full or there's no more work to do (when entering wait)
- Rework event loop sequencing w.r.t. checking for tasks/completions
- Change some longs to back to ints where I don't think they need to be
converted
- Track head/tail and masks (constant) as instance vars to avoid reading
from the mapped buffer every time

Result

Hopefully further perf gains

normanmaurer commented:
@njhill I think you missed updating the tests as this does not compile ;) Also did you run the test suite yet with the changes ?

njhill commented:
@normanmaurer ah, apologies I think I must have missed that when rebasing, will fix that

> Also did you run the test suite yet with the changes ?

Nope as I said above this isn't tested yet, I can do that soon after getting new kernel set up... intended right now more for discussion to see what you think of the changes/ideas

njhill committed:
Fix test compile errors

njhill commented:
@normanmaurer tests updated... again, I haven't run them will do so and debug when I get a chance

The main idea in the changes is deferring submissions and combining with the getevents (waiting) call.

njhill committed:
Fix completion head increment bug

normanmaurer commented:
@njhill so I didn't look very deeply yet but I can tell you with this basically all tests time out ;)

normanmaurer commented:
@njhill btw I like the "eventfd" change related to not using POLLIN for it. Like you said this should just work. Want to do this a a separate PR first so we can split this up a bit and so also make it easier to debug ?

njhill commented:
@normanmaurer sure, I can break some of this up. I'm hoping to have some time to test/debug later today

njhill committed:
Various fixes

1Jo1 commented:
@njhill thanks for you PR, FAST_POLL non blocking magic is unfortunately not possible anymore [io_uring ML](https://lore.kernel.org/io-uring/2616564c-aa35-1f25-c9de-4d6735e73a9b@kernel.dk/T/#t) btw I think performance gains is marginal as only 8 bytes will be read

> reserve an entry in the SQE array

 I'm not sure if understand it correctly, I do think that that the reverse an entry is not really necessary as each time you call an io_uring(wait) you flush the sqe entries in front of it

1Jo1 commented on transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:
```diff
@@ -87,30 +95,36 @@
         this.ringAddress = ringAddress;
         this.ringFd = ringFd;
 
+        this.ringMask = PlatformDependent.getInt(kRingMaskAddress);
```
I like your sqe changes 👍 it's more readable

1Jo1 commented:
btw you need a custom kernel https://git.kernel.org/pub/scm/linux/kernel/git/axboe/linux-block.git/log/?h=io_uring-5.9 for testing

1Jo1 commented on transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:
```diff
@@ -28,6 +28,11 @@
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(IOUringSubmissionQueue.class);
 
+    // this can be set smaller than the ring size to flush submissions more frequently
+    private static final int MAX_SUBMISSION_BATCH_SIZE = Integer.MAX_VALUE;
```
it's probably not good idea to use as default, if events are constantly coming in cqe loop(for example 200 etc., it will "never" be submitted, latency would be very bad

1Jo1 commented on transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:
```diff
@@ -59,8 +64,11 @@
-    private long sqeHead;
-    private long sqeTail;
+    private final int ringMask;
+    private int sqeSlots;
```
in general I like your sqe changes 👍 it's more readable

1Jo1 commented on transport-native-io_uring/src/main/java/io/netty/channel/uring/IOUringSubmissionQueue.java:
```diff
@@ -28,6 +28,11 @@
     private static final InternalLogger logger = InternalLoggerFactory.getInstance(IOUringSubmissionQueue.class);
 
+    // this can be set smaller than the ring size to flush submissions more frequently
+    private static final int MAX_SUBMISSION_BATCH_SIZE = Integer.MAX_VALUE;
```
it's probably not good idea to use as default value, if events are constantly coming in cqe loop(for example 200 etc., it will "never" be submitted, latency would be very bad

njhill replied:
Right I was thinking this could be something to experiment with, and maybe a different metric like EL iterations or nanotime.

But I actually don't think the example you gave would be very likely since there's likely to be a similar rate of submissions to completions in general.. after all they are 1-1.

njhill committed:
More fixes - get tests working

- Tidy up setData method, zero SQE array upfront
- Make eventfd non-blocking and handle channel shutdown race (like
epoll)

njhill commented:
Hey @1Jo1 @normanmaurer .. thanks for your review, somehow I didn't notice it until just now after I already pushed some updates.

I was testing with 5.9-rc3, I guess won't have all the latest fixes. The tests do all seem to work individually, but when running all together seem flaky, with weird errors like getting a null ring. I am running in a VM so might also be a resource use thing.

> FAST_POLL non blocking magic is unfortunately not possible anymore io_uring ML btw I think performance gains is marginal as only 8 bytes will be read

Right but that thread applies to non-blocking I think? It seems to work if eventfd is made blocking. I didn't think the advantage would be related to the data size but rather avoiding the poll+read roundtrip. I guess those could be linked but it would still result in an intermediate CQE.

njhill commented:
> I'm not sure if understand it correctly, I do think that that the reverse an entry is not really necessary as each time you call an io_uring(wait) you flush the sqe entries in front of it

Yeah I'm also not really sure about this part. I thought that io_uring_enter should always consume all SQEs submitted so would make sense to just reset to 0 but then saw that this doesn't seem to always be the case, not sure why. The thought was that since there's a "constant" SQE that's often being submitted it might be better to dedicate an array element to it and just write the index every time its needed. I guess the benefit could be negligible though.

njhill commented:
I will break these up into separate PRs now...

1Jo1 commented:
@njhill thanks

> I was testing with 5.9-rc3, I guess won't have all the latest fixes. The tests do all seem to work individually, but when running all together seem flaky, with weird errors like getting a null ring. I am running in a VM so might also be a resource use thing.

yeah 5.9-rc3 doesnt work, here I'm using this custom kernel https://github.com/1Jo1/linux/tree/io_uring-read-fix, (until the linux 5.9-rc4 will be released) did you increase memlock limit?


> Yeah I'm also not really sure about this part. I thought that io_uring_enter should always consume all SQEs submitted so would make sense to just reset to 0 but then saw that this doesn't seem to always be the case, not sure why. The thought was that since there's a "constant" SQE that's often being submitted it might be better to dedicate an array element to it and just write the index every time its needed. I guess the benefit could be negligible though.

Ok I got it, with io_uring_enter you always specify how many event you want to submit, so we dont need that

> I will break these up into separate PRs now...

I really like you sqe implementation, can you create a new PR? so I can merge it

normanmaurer commented:
Or you can also just clone this branch:

git://git.kernel.dk/linux-block 5.8-stable
